src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp

```diff
@@ -992,6 +1002,280 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_JoinerStart(voi
+        break;
+    case OT_ERROR_NOT_FOUND:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Domain name or RRset does not exist");
+        break;
+    case OT_ERROR_NOT_IMPLEMENTED:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Server does not support query type");
+        break;
+    case OT_ERROR_SECURITY:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Operation refused for security reasons");
+        break;
+    case OT_ERROR_DUPLICATED:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Domain name or RRset is duplicated");
+        break;
+    case OT_ERROR_RESPONSE_TIMEOUT:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Timed out waiting on server response");
+        break;
+    case OT_ERROR_INVALID_ARGS:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Invalid service structure detected");
+        break;
+    case OT_ERROR_NO_BUFS:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Insufficient buffer to handle message");
+        break;
+    case OT_ERROR_FAILED:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Internal server error occurred");
+        break;
+    default:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Unknown error occurred");
+        break;
+    }
+}
+
+template <class ImplClass>
+CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_AddSrpService(const char * aInstanceName, const char * aName,
+                                                                               uint16_t aPort, uint32_t aLeaseInterval,
+                                                                               uint32_t aKeyLeaseInterval)
+{
+    CHIP_ERROR error                         = CHIP_NO_ERROR;
+    typename SrpClient::Service * srpService = nullptr;
+
+    Impl()->LockThreadStack();
+
+    VerifyOrExit(aInstanceName, error = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(strlen(aInstanceName) < SrpClient::kMaxNameSize, error = CHIP_ERROR_INVALID_STRING_LENGTH);
+    VerifyOrExit(aName, error = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(strlen(aName) < SrpClient::kMaxNameSize, error = CHIP_ERROR_INVALID_STRING_LENGTH);
+
+    // Check if service with desired instance name already exists and try to find empty slot in array for new service
+    for (typename SrpClient::Service & service : mSrpClient.mServices)
+    {
+        if (strcmp(service.mInstanceName, "") == 0)
+        {
+            // assign first empty slot in array for a new service
+            srpService = srpService ? srpService : &service;
+        }
+        else
+        {
+            VerifyOrExit(strcmp(service.mInstanceName, aInstanceName) != 0, error = MapOpenThreadError(OT_ERROR_DUPLICATED));
+        }
+    }
+
+    // Verify is there an empty place for new service.
+    VerifyOrExit(srpService, error = MapOpenThreadError(OT_ERROR_NO_BUFS));
+
+    otSrpClientSetLeaseInterval(mOTInst, aLeaseInterval);
```

Comment:
Isn't there a way to configure the lease times per service? Should we check if all registered services don't have conflicting demands?

Reply:
AFAIK currently there is no way to do that. If you set lease time it applies to all services added after that operation, but it doesn't change anything for the ones added previously. That's why I decided to call otSrpClientSetLeaseInterval on adding each service, what allows user to call AddSrpService with different lease time value.